### PR TITLE
Fix assert in reduce_by_segment_impl

### PR DESCRIPTION
**Pull request by SergeyKopienko**

### CODE REVIEW

**Review by MikeDvorskiy — APPROVED** [review]
broken CI

File: include/oneapi/dpl/internal/reduce_by_segment_impl.h [hunk 1]
```diff
@@ -167,9 +167,9 @@ reduce_by_segment_impl(Policy&& policy, InputIterator1 first1, InputIterator1 la
     auto keep_values = __ranges::__get_sycl_range<__bknd::access_mode::read, InputIterator2>();
     auto value_buf = keep_values(first2, first2 + n);
     auto keep_key_outputs = __ranges::__get_sycl_range<__bknd::access_mode::write, OutputIterator1>();
-    auto key_output_buf = keep_key_outputs(result1, result1 + n);
+    auto key_output_buf = keep_key_outputs(result1, dpl::end(result1.get_buffer()));
```
Comment (MikeDvorskiy) [inline include/oneapi/dpl/internal/reduce_by_segment_impl.h:170]: result1, result2 can be a USM pointer, host pointer or host iterator....

**Review by MikeDvorskiy — CHANGES_REQUESTED** [review]

File: include/oneapi/dpl/internal/reduce_by_segment_impl.h (hunk 1 above)
Comment (SergeyKopienko) [reply include/oneapi/dpl/internal/reduce_by_segment_impl.h]: Fixed

File: include/oneapi/dpl/pstl/hetero/dpcpp/utils_ranges_sycl.h [hunk 2]
```diff
@@ -522,10 +522,16 @@ struct __get_sycl_range
     {
         assert(__first < __last);
         using value_type = val_t<_Iter>;
+
+        auto __offset = __first - oneapi::dpl::begin(__first.get_buffer());
+        auto __n = __last - __first;
+        auto __size = __first.get_buffer().size();
+        if (__offset + __n > __size)
+            __n = __size - __offset;
```
Comment (MikeDvorskiy) [inline include/oneapi/dpl/pstl/hetero/dpcpp/utils_ranges_sycl.h:530]: I believe we don't need to "correct" size of sequence.. We should an additional assert here like `assert(__offset + __n <= __size)`.

Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/utils_ranges_sycl.h]: Fixed: changed to assert; fixed error in histogram test.

File: examples/histogram/src/main.cpp [hunk 3]
```diff
@@ -68,25 +68,30 @@ void sparse_histogram(std::vector<uint64_t> &input) {
...
-  oneapi::dpl::reduce_by_segment(
+  const auto result = oneapi::dpl::reduce_by_segment(
       oneapi::dpl::execution::dpcpp_default, oneapi::dpl::begin(histogram_buf),
       oneapi::dpl::end(histogram_buf), oneapi::dpl::begin(_const_buf),
-      oneapi::dpl::begin(histogram_values_buf),
-      oneapi::dpl::begin(histogram_counts_buf));
+      histogram_values_buf_begin,
+      histogram_counts_buf_begin);
+
+  assert(result.first - histogram_values_buf_begin == result.second - histogram_counts_buf_begin);
```
Comment (MikeDvorskiy) [inline examples/histogram/src/main.cpp:92]: I would move the line after `const auto num_bins =...` 

`assert(num_bins == result.second - histogram_counts_buf_begin);`

**Review by MikeDvorskiy — CHANGES_REQUESTED** [review]

**Review by MikeDvorskiy — APPROVED** [review]
LGTM

**Comment by SergeyKopienko** [comment]
Added assertion on buffer size check (a couple of iterators to a range).
Fixed a buffer size error in histogram example.

**Pull request merged**